src/config/config.rs

```diff
@@ -139,17 +157,185 @@ pub struct CSIConfig {
     pub worker_port: u16,
 }
 
+impl Config {
+    /// Load config from command line, config file and environment variables
+    ///
+    /// Priority: command line > config file > environment variables
+    #[inline]
+    pub fn load() -> Result<Self, DatenLordError> {
+        // Try to load from command line
+        let arg_conf = Self::parse();
+
+        let mut builder: serfig::Builder<Self> = serfig::Builder::default();
+
+        // Load from config file
+        {
+            let config_file = match arg_conf.config_file.clone() {
+                Some(path) => path,
+                None => {
+                    if let Ok(path) = env::var("CONFIG_FILE") {
+                        path
+                    } else {
+                        String::new()
+                    }
+                }
+            };
+
+            if !config_file.is_empty() {
+                builder = builder.collect(from_file(Toml, &config_file));
+            }
+        }
+
+        // Load from environment
+        builder = builder.collect(from_env());
+
+        // Load from arguments
+        builder = builder.collect(from_self(arg_conf));
+
+        // Build the config with multi sources
+        let conf = builder.build()?;
+
+        // Check valid
+        match conf.check_valid() {
+            Ok(true) => Ok(conf),
+            Ok(false) => Err(DatenLordError::ArgumentInvalid { context: vec![] }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Check the config values are valid
+    #[inline]
+    pub fn check_valid(&self) -> Result<bool, DatenLordError> {
+        // Check role/node_name/node_ip/mount_path
+        let checked_values = [&self.role, &self.node_name, &self.node_ip, &self.mount_path];
+
+        let valid = checked_values.iter().all(|x| x.is_some());
+
+        Ok(valid)
+    }
+}
+
 #[cfg(test)]
 #[allow(clippy::unwrap_used)]
 mod tests {
     use std::net::IpAddr;
     use std::num::NonZeroUsize;
     use std::str::FromStr;
 
+    use std::env;
+    use std::io::Write;
+    use tempfile::NamedTempFile;
+
     use super::*;
     use crate::config::inner::{InnerConfig, Role, StorageParams as InnerStorageParams};
     use crate::config::SoftLimit;
 
+    #[test]
+    fn test_load_config_from_file() {
+        let mut temp_file = NamedTempFile::new().unwrap();
+        let temp_file_path = temp_file.path().to_owned();
+        writeln!(
+            temp_file,
+            r#"
+            role = "node"
+            node_name = "node1"
+            node_ip = "127.0.0.1"
+            server_port = 8800
+            mount_path = "/tmp/datenlord_data_dir"
+            kv_server_list = [
+              "127.0.0.1:2379"
+            ]
+        "#
+        )
+        .unwrap();
+
+        let mut builder: serfig::Builder<Config> = serfig::Builder::default();
+        builder = builder.collect(from_file(Toml, temp_file_path.to_str().unwrap()));
+        let conf = builder.build().unwrap();
```

Comment:
You seems to be writing UT for `serfig` crate, but not for `datenlord`. Perhaps you should refactor `Config::load()` to make it possible to test your own logic. In fact, the `Config::load()` method is not coverred by test cases at all.

Reply:
Ok, I try to load config from basic `config::parse()`, and it can be integrated with unit test.

Some mock operation may be not elegant in unit test.
> https://www.reddit.com/r/rust/comments/xig8xm/best_method_for_cli_testing_with_clap/